op-node/rollup/derive/engine_consolidate.go

```diff
@@ -26,6 +28,27 @@ func AttributesMatchBlock(attrs *eth.PayloadAttributes, parentHash common.Hash,
 	}
 	for i, otx := range attrs.Transactions {
 		if expect := block.Transactions[i]; !bytes.Equal(otx, expect) {
+			if i == 0 {
+				var safeTx, unsafeTx types.Transaction
+				var safeInfo, unsafeInfo L1BlockInfo
+				errSafe := (&safeTx).UnmarshalBinary(otx)
+				errUnsafe := (&unsafeTx).UnmarshalBinary(block.Transactions[i])
+				if errSafe == nil && errUnsafe == nil {
+					errSafe = (&safeInfo).UnmarshalBinary(safeTx.Data())
+					errUnsafe = (&unsafeInfo).UnmarshalBinary(unsafeTx.Data())
+					if errSafe == nil && errUnsafe == nil {
+						l.Warn("L1 Info transaction differs", "number", uint64(block.BlockNumber), "time", uint64(block.Timestamp),
+							"safe_l1_number", safeInfo.Number, "safe_l1_hash", safeInfo.BlockHash,
+							"safe_l1_time", safeInfo.Time, "safe_seq_num", safeInfo.SequenceNumber,
+							"unsafe_l1_number", unsafeInfo.Number, "unsafe_l1_hash", unsafeInfo.BlockHash,
+							"unsafe_l1_time", unsafeInfo.Time, "unsafe_seq_num", unsafeInfo.SequenceNumber)
+					} else {
+						l.Warn("failed to umarshal l1 info", "errSafe", errSafe, "errUnsafe", errUnsafe)
+					}
+				} else {
+					l.Warn("failed to umarshal tx", "errSafe", errSafe, "errUnsafe", errUnsafe)
+				}
```

Comment:
nit: Would it be worth extracting this as a separate method? Then the main business logic would be easier to read, but also you could probably avoid the deep nesting here by returning early on errors.

But by all means ignore me, my experience with styling go code involves watching about three YouTube videos. :)

Reply:
good structure is good structure regardless of language. I can pull this out into it's own method - that'll help a bunch. We really shouldn't hit this error case, but I've had to recreate this log enough that I finally wanted to merge it.